### ISSUE TITLE
KMeans clusters 

### ISSUE THREAD

**Issue by nixeimar** [oracle/tribuo#268]
Hi, 

I am using KMeans clustering and I want to plot the different clusters created by KMeans along with the corresponding data to each cluster but I cannot find a way to do that. How can I have access to the clusters of the model?

I am using Tribuo 4.2. 

Thank you in advance. 

Best regards, 
Nikos

**Comment by nixeimar** [comment]
Thank you again for your reply! I will work with it.

**Comment by nixeimar** [comment]
I do not know if I set this as a bug but here what I have done:

I created this small code where I read a dataset with CVSLoader and I use KMeans to compute the clusters and I get multiple zeros for centroids. Can  you please take a look and comment on this? I check
[concrete.csv](https://github.com/oracle/tribuo/files/9497217/concrete.csv)
ed also with R and python and there should not be any zeros. Thank you again! 

Here is the code:

Path csvPath = Paths.get("C:/Users/nikol/Desktop/","concrete.csv");
CSVLoader csvLoader = new CSVLoader<>(',', new LabelFactory() );
MutableDataset dataset = new MutableDataset<>(csvLoader.loadDataSource( csvPath, "Cement"));

KMeansTrainer trainer = new KMeansTrainer(10, /* centroids */
				200, /* iterations */
				Distance.EUCLIDEAN, /* distance function */
				1, /* number of compute threads */
				1 /* RNG seed */
				);
		
KMeansModel model = trainer.train(dataset);

List<List<Feature>> centroids = model.getCentroids();
ImmutableFeatureMap out = model.getFeatureIDMap();
for (List<Feature> centroid : centroids) {
	System.out.println(centroid);
}


I have attached the dataset and  the output that I get:

[(Age, 44,449541), (BlastFurnaceSlag, 24,783180), (CoarseAggregate, 925,746483), (CompressiveStrength, 34,370917), (FineAggregate, 844,935780), (FlyAsh, 66,897554), (Superplasticizer, 7,719266), (Water, 176,907951)]
[(Age, 0,000000), (BlastFurnaceSlag, 0,000000), (CoarseAggregate, 0,000000), (CompressiveStrength, 0,000000), (FineAggregate, 0,000000), (FlyAsh, 0,000000), (Superplasticizer, 0,000000), (Water, 0,000000)]
[(Age, 52,893151), (BlastFurnaceSlag, 170,927945), (CoarseAggregate, 943,610137), (CompressiveStrength, 38,615123), (FineAggregate, 729,920000), (FlyAsh, 29,472603), (Superplasticizer, 5,677260), (Water, 190,072329)]
[(Age, 0,000000), (BlastFurnaceSlag, 0,000000), (CoarseAggregate, 0,000000), (CompressiveStrength, 0,000000), (FineAggregate, 0,000000), (FlyAsh, 0,000000), (Superplasticizer, 0,000000), (Water, 0,000000)]
[(Age, 39,026627), (BlastFurnaceSlag, 16,626923), (CoarseAggregate, 1050,206213), (CompressiveStrength, 34,197308), (FineAggregate, 751,695562), (FlyAsh, 68,582840), (Superplasticizer, 5,308876), (Water, 176,890533)]
[(Age, 0,000000), (BlastFurnaceSlag, 0,000000), (CoarseAggregate, 0,000000), (CompressiveStrength, 0,000000), (FineAggregate, 0,000000), (FlyAsh, 0,000000), (Superplasticizer, 0,000000), (Water, 0,000000)]
[(Age, 0,000000), (BlastFurnaceSlag, 0,000000), (CoarseAggregate, 0,000000), (CompressiveStrength, 0,000000), (FineAggregate, 0,000000), (FlyAsh, 0,000000), (Superplasticizer, 0,000000), (Water, 0,000000)]
[(Age, 0,000000), (BlastFurnaceSlag, 0,000000), (CoarseAggregate, 0,000000), (CompressiveStrength, 0,000000), (FineAggregate, 0,000000), (FlyAsh, 0,000000), (Superplasticizer, 0,000000), (Water, 0,000000)]
[(Age, 0,000000), (BlastFurnaceSlag, 0,000000), (CoarseAggregate, 0,000000), (CompressiveStrength, 0,000000), (FineAggregate, 0,000000), (FlyAsh, 0,000000), (Superplasticizer, 0,000000), (Water, 0,000000)]
[(Age, 0,000000), (BlastFurnaceSlag, 0,000000), (CoarseAggregate, 0,000000), (CompressiveStrength, 0,000000), (FineAggregate, 0,000000), (FlyAsh, 0,000000), (Superplasticizer, 0,000000), (Water, 0,000000)]
[concrete.csv](https://github.com/oracle/tribuo/files/9497222/concrete.csv)

**Comment by nixeimar** [comment]
Thank you again for your reply! Changing to KMeans++ did the trick. There are no zeros appearing now and the results are as expected.

**Comment by Craigacp** [comment]
Ok. Feel free to re-open the issue if you have more issues with KMeans.
